src/compiler/rust/canister_methods/src/upload_file_chunk/mod.rs

```diff
@@ -0,0 +1,80 @@
+pub mod versions;
+pub mod write_chunk;
+
+use quote::quote;
+
+use crate::hash_file;
+
+pub fn get_upload_file_chunk() -> proc_macro2::TokenStream {
+    let write_chunk_src = write_chunk::get_write_chunk();
+    let check_if_latest_version_src = versions::get_check_if_latest_version_src();
+    let hash_file_src = hash_file::get_hash_file();
+
+    quote! {
+        #write_chunk_src
+        #check_if_latest_version_src
+        #hash_file_src
+
+        #[ic_cdk_macros::update]
+        pub fn upload_file_chunk(
+            dest_path: String,
+            timestamp: u64,
+            start_index: u64,
+            file_bytes: Vec<u8>,
+            total_file_len: u64,
+        ) {
+            if !ic_cdk::api::is_controller(&ic_cdk::api::caller()) {
```

Comment:
Let's use a guard, or at least create a function that has this logic in it and use that everywhere

Reply:
done